panic on wallet connect

Running btcd on commit 1d60e5dba5dfe990c12a78c56d95ae5970d76466.
When btcwallet connects...

<pre>
10:39:47 2014-02-10 [INF] RPCS: RPC server listening on 127.0.0.1:18334
10:39:48 2014-02-10 [INF] AMGR: Loaded 7205 addresses from '/home/dhill/.btcd/data/testnet/peers.json'
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x8 pc=0x447782]

goroutine 36 [running]:
runtime.panic(0xb5bfa0, 0x1632008)
        /home/dhill/go/src/pkg/runtime/panic.c:250 +0xb4
main.(*wsContext).AddTxRequest(0xc21093f640, 0xc2108b4540, 0xc2103ef5a0, 0x22)
        /home/dhill/.go/src/github.com/conformal/btcd/rpcwebsocket.go:112 +0x1f2
main.handleNotifyNewTXs(0xc2112f23f0, 0x7f58b5daa010, 0xc2103ddd50, 0xc2108b4540, 0xc2108b4600, ...)
        /home/dhill/.go/src/github.com/conformal/btcd/rpcwebsocket.go:379 +0x59f
main.respondToAnyCmd(0x7f58b5daa010, 0xc2103ddd50, 0xc2112f23f0, 0xc2108b4540, 0xc2108b4600, ...)
        /home/dhill/.go/src/github.com/conformal/btcd/rpcwebsocket.go:306 +0x1cc
main.func·028()
        /home/dhill/.go/src/github.com/conformal/btcd/rpcwebsocket.go:812 +0xd4
created by main.(*rpcServer).websocketJSONHandler
        /home/dhill/.go/src/github.com/conformal/btcd/rpcwebsocket.go:821 +0x3d8

goroutine 16 [chan receive]:
main.btcdMain(0x0, 0x0, 0x0)
        /home/dhill/.go/src/github.com/conformal/btcd/btcd.go:123 +0xc09
main.main()
        /home/dhill/.go/src/github.com/conformal/btcd/btcd.go:152 +0x67

goroutine 18 [syscall]:
os/signal.loop()
        /home/dhill/go/src/pkg/os/signal/signal_unix.go:21 +0x2c
created by os/signal.init·1
        /home/dhill/go/src/pkg/os/signal/signal_unix.go:27 +0x3f

goroutine 19 [semacquire]:
sync.runtime_Syncsemacquire(0xc210078090)
        /home/dhill/go/src/pkg/runtime/sema.goc:257 +0xc0
sync.(*Cond).Wait(0xc210078080)
        /home/dhill/go/src/pkg/sync/cond.go:62 +0xa1
github.com/conformal/seelog.(*asyncLoopLogger).processItem(0xc21000a5a0, 0x0)
        /home/dhill/.go/src/github.com/conformal/seelog/behavior_asynclooplogger.go:50 +0x147
github.com/conformal/seelog.(*asyncLoopLogger).processQueue(0xc21000a5a0)
        /home/dhill/.go/src/github.com/conformal/seelog/behavior_asynclooplogger.go:63 +0x52
created by github.com/conformal/seelog.newAsyncLoopLogger
        /home/dhill/.go/src/github.com/conformal/seelog/behavior_asynclooplogger.go:40 +0xba
goroutine 20 [semacquire]:
sync.runtime_Syncsemacquire(0xc210078210)
        /home/dhill/go/src/pkg/runtime/sema.goc:257 +0xc0
sync.(*Cond).Wait(0xc210078200)
        /home/dhill/go/src/pkg/sync/cond.go:62 +0xa1
github.com/conformal/seelog.(*asyncLoopLogger).processItem(0xc21000a660, 0x0)
        /home/dhill/.go/src/github.com/conformal/seelog/behavior_asynclooplogger.go:50 +0x147
github.com/conformal/seelog.(*asyncLoopLogger).processQueue(0xc21000a660)
        /home/dhill/.go/src/github.com/conformal/seelog/behavior_asynclooplogger.go:63 +0x52
created by github.com/conformal/seelog.newAsyncLoopLogger
        /home/dhill/.go/src/github.com/conformal/seelog/behavior_asynclooplogger.go:40 +0xba

goroutine 81 [select]:
main.(*AddrManager).addressHandler(0xc2100e3800)
        /home/dhill/.go/src/github.com/conformal/btcd/addrmanager.go:393 +0x234
created by main.(*AddrManager).Start
        /home/dhill/.go/src/github.com/conformal/btcd/addrmanager.go:626 +0x183

goroutine 17 [syscall]:
runtime.goexit()
        /home/dhill/go/src/pkg/runtime/proc.c:1438

goroutine 22 [chan receive]:
github.com/conformal/seelog.(*asyncAdaptiveLogger).processQueue(0xc210071150)
        /home/dhill/.go/src/github.com/conformal/seelog/behavior_adaptivelogger.go:127 +0xad
created by github.com/conformal/seelog.newAsyncAdaptiveLogger
        /home/dhill/.go/src/github.com/conformal/seelog/behavior_adaptivelogger.go:84 +0x6c9

goroutine 23 [select]:
github.com/conformal/goleveldb/leveldb.(*DB).compactionError(0xc210103000)
        /home/dhill/.go/src/github.com/conformal/goleveldb/leveldb/db_compaction.go:115 +0x24f
created by github.com/conformal/goleveldb/leveldb.openDB
        /home/dhill/.go/src/github.com/conformal/goleveldb/leveldb/db.go:114 +0x5e0

goroutine 24 [runnable]:
github.com/conformal/goleveldb/leveldb.func·014(0x0, 0x0)
        /home/dhill/.go/src/github.com/conformal/goleveldb/leveldb/db_compaction.go:287 +0x696
github.com/conformal/goleveldb/leveldb.(*DB).transact(0xc210103000, 0xc1caf0, 0xb, 0x7f58bc23bd88, 0x7f58bc23bb48)
        /home/dhill/.go/src/github.com/conformal/goleveldb/leveldb/db_compaction.go:157 +0x3ae
github.com/conformal/goleveldb/leveldb.(*DB).doCompaction(0xc210103000, 0xc21004f1c0, 0x7f58bc23be00)
        /home/dhill/.go/src/github.com/conformal/goleveldb/leveldb/db_compaction.go:407 +0x1475
github.com/conformal/goleveldb/leveldb.(*DB).compaction(0xc210103000)
        /home/dhill/.go/src/github.com/conformal/goleveldb/leveldb/db_compaction.go:472 +0x29d
created by github.com/conformal/goleveldb/leveldb.openDB
        /home/dhill/.go/src/github.com/conformal/goleveldb/leveldb/db.go:117 +0x621

goroutine 25 [select]:
github.com/conformal/goleveldb/leveldb.(*DB).writeJournal(0xc210103000)
        /home/dhill/.go/src/github.com/conformal/goleveldb/leveldb/db_write.go:36 +0x1aa
created by github.com/conformal/goleveldb/leveldb.openDB
        /home/dhill/.go/src/github.com/conformal/goleveldb/leveldb/db.go:118 +0x638

goroutine 26 [select]:
main.mainInterruptHandler()
        /home/dhill/.go/src/github.com/conformal/btcd/signal.go:28 +0x355
created by main.addInterruptHandler
        /home/dhill/.go/src/github.com/conformal/btcd/signal.go:55 +0x145

goroutine 27 [IO wait]:
net.runtime_pollWait(0x7f58bfc14138, 0x72, 0x0)
        /home/dhill/go/src/pkg/runtime/netpoll.goc:138 +0x5d
net.(*pollDesc).Wait(0xc210071840, 0x72, 0x7f58bfc09288, 0xb)
        /home/dhill/go/src/pkg/net/fd_poll_runtime.go:81 +0x50
net.(*pollDesc).WaitRead(0xc210071840, 0xb, 0x7f58bfc09288)
        /home/dhill/go/src/pkg/net/fd_poll_runtime.go:86 +0x3e
net.(*netFD).accept(0xc2100717e0, 0xd1d608, 0x0, 0x7f58bfc09288, 0xb)
        /home/dhill/go/src/pkg/net/fd_unix.go:392 +0x426
net.(*TCPListener).AcceptTCP(0xc2100eb2c0, 0x209f4c8, 0x57ae82, 0xc2100f0aa8)
        /home/dhill/go/src/pkg/net/tcpsock_posix.go:233 +0x75
net.(*TCPListener).Accept(0xc2100eb2c0, 0x0, 0x16, 0xc210964010, 0x1)
        /home/dhill/go/src/pkg/net/tcpsock_posix.go:243 +0x35
main.(*server).listenHandler(0xc2100f0a80, 0x7f58bfc13170, 0xc2100eb2c0)
        /home/dhill/.go/src/github.com/conformal/btcd/server.go:416 +0x18a
created by main.(*server).Start
        /home/dhill/.go/src/github.com/conformal/btcd/server.go:720 +0x20a

goroutine 28 [IO wait]:
net.runtime_pollWait(0x7f58bfc14090, 0x72, 0x0)
        /home/dhill/go/src/pkg/runtime/netpoll.goc:138 +0x5d
net.(*pollDesc).Wait(0xc2100718b0, 0x72, 0x7f58bfc09288, 0xb)
        /home/dhill/go/src/pkg/net/fd_poll_runtime.go:81 +0x50
net.(*pollDesc).WaitRead(0xc2100718b0, 0xb, 0x7f58bfc09288)
        /home/dhill/go/src/pkg/net/fd_poll_runtime.go:86 +0x3e
net.(*netFD).accept(0xc210071850, 0xd1d608, 0x0, 0x7f58bfc09288, 0xb)
        /home/dhill/go/src/pkg/net/fd_unix.go:392 +0x426
net.(*TCPListener).AcceptTCP(0xc2100eb2b0, 0x20c04f8, 0x57ae82, 0xc2100f0aa8)
        /home/dhill/go/src/pkg/net/tcpsock_posix.go:233 +0x75
net.(*TCPListener).Accept(0xc2100eb2b0, 0x0, 0x16, 0xc210964020, 0x1)
        /home/dhill/go/src/pkg/net/tcpsock_posix.go:243 +0x35
main.(*server).listenHandler(0xc2100f0a80, 0x7f58bfc13170, 0xc2100eb2b0)
        /home/dhill/.go/src/github.com/conformal/btcd/server.go:416 +0x18a
created by main.(*server).Start
        /home/dhill/.go/src/github.com/conformal/btcd/server.go:720 +0x20a

goroutine 29 [syscall]:
net._C2func_getaddrinfo(0x7f58ac1333d0, 0x0, 0xc211499630, 0xc21161a778, 0x986020, ...)
        /tmp/go-build570675839/net/_obj/_cgo_defun.c:50 +0x36
net.cgoLookupIPCNAME(0xc96ad0, 0x22, 0x0, 0x0, 0x0, ...)
        /home/dhill/go/src/pkg/net/cgo_unix.go:96 +0x1ed
net.cgoLookupIP(0xc96ad0, 0x22, 0x0, 0x0, 0x0, ...)
        /home/dhill/go/src/pkg/net/cgo_unix.go:148 +0x77
net.lookupIP(0xc96ad0, 0x22, 0x0, 0x0, 0x0, ...)
        /home/dhill/go/src/pkg/net/lookup_unix.go:64 +0x71
net.func·022(0x164c690, 0xc210f9cde0, 0x7f58bf436b30, 0x7f58bf436b10)
        /home/dhill/go/src/pkg/net/lookup.go:41 +0x4e
net.(*singleflight).Do(0x164c690, 0xc96ad0, 0x22, 0x7f58bf436b80, 0x0, ...)
        /home/dhill/go/src/pkg/net/singleflight.go:45 +0x29c
net.lookupIPMerge(0xc96ad0, 0x22, 0x0, 0x0, 0x0, ...)
        /home/dhill/go/src/pkg/net/lookup.go:42 +0xef
net.LookupIP(0xc96ad0, 0x22, 0x0, 0x0, 0x0, ...)
        /home/dhill/go/src/pkg/net/lookup.go:31 +0x71
main.btcdLookup(0xc96ad0, 0x22, 0x1e, 0xc21161d3c0, 0x1, ...)
        /home/dhill/.go/src/github.com/conformal/btcd/config.go:582 +0x1b8
main.dnsDiscover(0xc96ad0, 0x22, 0x986060, 0x7f58bf436d10, 0x495e05)
        /home/dhill/.go/src/github.com/conformal/btcd/discovery.go:133 +0x134
main.(*server).seedFromDNS(0xc2100f0a80)
        /home/dhill/.go/src/github.com/conformal/btcd/server.go:439 +0x118
main.(*server).peerHandler(0xc2100f0a80)
        /home/dhill/.go/src/github.com/conformal/btcd/server.go:494 +0x330
created by main.(*server).Start
        /home/dhill/.go/src/github.com/conformal/btcd/server.go:726 +0x26c

goroutine 30 [IO wait]:
net.runtime_pollWait(0x7f58bfc13fe8, 0x72, 0x0)
        /home/dhill/go/src/pkg/runtime/netpoll.goc:138 +0x5d
net.(*pollDesc).Wait(0xc21104af40, 0x72, 0x7f58bfc09288, 0xb)
        /home/dhill/go/src/pkg/net/fd_poll_runtime.go:81 +0x50
net.(*pollDesc).WaitRead(0xc21104af40, 0xb, 0x7f58bfc09288)
        /home/dhill/go/src/pkg/net/fd_poll_runtime.go:86 +0x3e
net.(*netFD).accept(0xc21104aee0, 0xd1d608, 0x0, 0x7f58bfc09288, 0xb)
        /home/dhill/go/src/pkg/net/fd_unix.go:392 +0x426
net.(*TCPListener).AcceptTCP(0xc21148b2a8, 0x2102708, 0xc21132bb20, 0x6fcf09)
        /home/dhill/go/src/pkg/net/tcpsock_posix.go:233 +0x75
net.(*TCPListener).Accept(0xc21148b2a8, 0x485663, 0xd1d630, 0x7f58bf434e78, 0x8)
        /home/dhill/go/src/pkg/net/tcpsock_posix.go:243 +0x35
crypto/tls.(*listener).Accept(0xc21132bb20, 0x0, 0x0, 0x0, 0x0)
        /home/dhill/go/src/pkg/crypto/tls/tls.go:46 +0x6f
net/http.(*Server).Serve(0xc21132c050, 0x7f58b5da8230, 0xc21132bb20, 0x0, 0x0)
        /home/dhill/go/src/pkg/net/http/server.go:1624 +0xa2
main.func·026(0x7f58b5da8230, 0xc21132bb20)
        /home/dhill/.go/src/github.com/conformal/btcd/rpcserver.go:195 +0x170
created by main.(*rpcServer).Start
        /home/dhill/.go/src/github.com/conformal/btcd/rpcserver.go:198 +0x52f

goroutine 31 [semacquire]:
sync.runtime_Semacquire(0xc21148b2c0)
        /home/dhill/go/src/pkg/runtime/sema.goc:199 +0x30
sync.(*WaitGroup).Wait(0xc2100f0b18)
        /home/dhill/go/src/pkg/sync/waitgroup.go:127 +0x1c1
main.(*server).WaitForShutdown(0xc2100f0a80)
        /home/dhill/.go/src/github.com/conformal/btcd/server.go:770 +0x44
main.func·005()
        /home/dhill/.go/src/github.com/conformal/btcd/btcd.go:116 +0x4a
created by main.btcdMain
        /home/dhill/.go/src/github.com/conformal/btcd/btcd.go:119 +0xbd3

goroutine 82 [select]:
main.(*blockManager).blockHandler(0xc2100f3b40)
        /home/dhill/.go/src/github.com/conformal/btcd/blockmanager.go:885 +0x8fe
created by main.(*blockManager).Start
        /home/dhill/.go/src/github.com/conformal/btcd/blockmanager.go:1108 +0x16f
</pre>

--- a/rpcwebsocket.go
+++ b/rpcwebsocket.go
@@ -80,7 +80,10 @@ func (r *wsContext) AddBlockUpdateRequest(n ntfnChan) {
 	r.Lock()
 	defer r.Unlock()
 
-	rc := r.connections[n]
+	rc, ok := r.connections[n]
+	if !ok {
+		return
+	}
 	rc.blockUpdates = true
 }
 
@@ -90,7 +93,10 @@ func (r *wsContext) AddAllNewTxRequest(n ntfnChan, verbose bool) {
 	r.Lock()
 	defer r.Unlock()
 
-	rc := r.connections[n]
+	rc, ok := r.connections[n]
+	if !ok {
+		return
+	}
 	rc.allTxUpdates = true
 	rc.verboseTxUpdates = verbose
 }
@@ -100,6 +106,12 @@ func (r *wsContext) AddTxRequest(n ntfnChan, addr string) {
 	r.Lock()
 	defer r.Unlock()
 
+	rc, ok := r.connections[n]
+	if !ok {
+		return
+	}
+	rc.txRequests[addr] = struct{}{}
+
 	clist, ok := r.txNotifications[addr]
 	if !ok {
 		clist = list.New()
@@ -107,9 +119,6 @@ func (r *wsContext) AddTxRequest(n ntfnChan, addr string) {
 	}
 
 	clist.PushBack(n)
-
-	rc := r.connections[n]
-	rc.txRequests[addr] = struct{}{}
 }
 
 func (r *wsContext) removeGlobalTxRequest(n ntfnChan, addr string) {
@@ -135,15 +144,18 @@ func (r *wsContext) AddSpentRequest(n ntfnChan, op *btcwire.OutPoint) {
 	r.Lock()
 	defer r.Unlock()
 
+	rc, ok := r.connections[n]
+	if !ok {
+		return
+	}
+	rc.spentRequests[*op] = struct{}{}
+
 	clist, ok := r.spentNotifications[*op]
 	if !ok {
 		clist = list.New()
 		r.spentNotifications[*op] = clist
 	}
 	clist.PushBack(n)
-
-	rc := r.connections[n]
-	rc.spentRequests[*op] = struct{}{}
 }
 
 func (r *wsContext) removeGlobalSpentRequest(n ntfnChan, op *btcwire.OutPoint) {
@@ -170,7 +182,10 @@ func (r *wsContext) RemoveSpentRequest(n ntfnChan, op *btcwire.OutPoint) {
 	defer r.Unlock()
 
 	r.removeGlobalSpentRequest(n, op)
-	rc := r.connections[n]
+	rc, ok := r.connections[n]
+	if !ok {
+		return
+	}
 	delete(rc.spentRequests, *op)
 }
 
@@ -180,15 +195,18 @@ func (r *wsContext) AddMinedTxRequest(n ntfnChan, txID *btcwire.ShaHash) {
 	r.Lock()
 	defer r.Unlock()
 
+	rc, ok := r.connections[n]
+	if !ok {
+		return
+	}
+	rc.minedTxRequests[*txID] = struct{}{}
+
 	clist, ok := r.minedTxNotifications[*txID]
 	if !ok {
 		clist = list.New()
 		r.minedTxNotifications[*txID] = clist
 	}
 	clist.PushBack(n)
-
-	rc := r.connections[n]
-	rc.minedTxRequests[*txID] = struct{}{}
 }
 
 func (r *wsContext) removeGlobalMinedTxRequest(n ntfnChan, txID *btcwire.ShaHash) {
@@ -221,7 +239,10 @@ func (r *wsContext) RemoveMinedTxRequest(n ntfnChan, txID *btcwire.ShaHash) {
 // mined transaction without grabbing any locks.
 func (r *wsContext) removeMinedTxRequest(n ntfnChan, txID *btcwire.ShaHash) {
 	r.removeGlobalMinedTxRequest(n, txID)
-	rc := r.connections[n]
+	rc, ok := r.connections[n]
+	if !ok {
+		return
+	}
 	delete(rc.minedTxRequests, *txID)
 }
 
@@ -780,6 +801,10 @@ func (s *rpcServer) websocketJSONHandler(r chan *btcjson.Reply, c handlerChans, 
 		// connection is not already authenticated.
 		s.ws.Lock()
 		rc := s.ws.connections[c.n]
+		// Note: c.n is guaranteed to be in the s.ws.connections map since it is
+		// called from the main walletReqNotifications for loop.  Any calls
+		// from that goroutine will always have valid connections since they
+		// are not removed till after that exits.
 		if _, ok := cmd.(*btcws.AuthenticateCmd); ok {
 			// Validate the provided credentials.
 			err := websocketAuthenticate(cmd, rc, s.authsha[:])

